DIPs/1NNN-ADR.md

```diff
@@ -0,0 +1,479 @@
+
+### Justifications
+
+The library wrapper may seem superfluous, however it serves three key roles:
+
+1. It divorces the compiler entirely from details of generated format strings.
+2. It provides necessary error checking capabilities.
+3. It provides an additional API for user functions to introspect the string, building on D's existing compile-time capabilities.
+
+The previous version of this DIP proposed a simple string literal instead of the result of `_d_interpolated_string` as the first argument passed. This is problematic because it introduces potentially subtle errors in likely usage.
+
+#### Wrong-use in unrelated function
+
+Consider a function in D today:
+
+```
+/++
+	Creates a window with the given title, width, and height. If
+	width or height are set to zero (the default), your window will
+	be automatically sized.
++/
+Window createWindow(string title, int width = 0, int height = 0);
+```
+
+A user has some code that calls it like so:
+
+```
+import std.conv;
+auto window = createWindow("Process debugger " ~ to!string(pid));
+```
+
+A new version of the compiler is released and the user, eager to try D's new string interpolation feature, rewrites that code as would be common in many other languages:
+
+```
+// look at this clean new code :D
+auto window = createWindow(i"Process debugger $pid");
+```
+
+It compiles without error. In bliss, the happy user thinks: "d rox". Then... the program is run and the window width is extraordinary. But the documentation says it will be automatically sized by default. The confused user wonders: why did this change? They run the program again and get another random width. Is the library incompatible with the new dmd version?
+
+After some time, they notice the window title has changed to "Process debugger %" before being cut off by the taskbar width. "This new dmd version must be super buggy," they think, "it corrupted my string!"
+
+Exasperated, they go online to ask the D gurus if anyone else has experienced this bizarre behavior. After an hour of wasted time, they finally mention having experimented with D's new string interpolation feature earlier in the day.
+
+The helpful people in the D forum point them at the spec: "I know it looks like a string," the Master explains, "but it is actually lowered to an auto-expanding tuple whose first element is a string. That's causing your problem. I'd love to fix it in the library, but it is literally impossible to differentiate this case from legitimate usage :("
+
+"D is weird," the user replies, deflated and defeated, "I think I'm just going to use Python. Have you see its f-strings?"
+
+***
+
+On the other hand, with `_d_interpolated_string`, it IS possible to tell those uses apart! In fact, the library author doesn't have to do anything - the user will see a helpful error message from the compiler and perhaps try `i"Process debugger $pid".idup` instead, or another suitable alternative.
```

Comment:
How? Be explictly descriptive here. Example would have assist your case here.

Reply:
It is just ordinary function overloading at this point; line 261 prior to the quoted segment (see https://github.com/dlang/DIPs/pull/186/files#diff-2d3c5bf5c5d1f001279a15e3449b2338R261 ) actually shows a sample `writefln` overload to specialize on one of these.

I could explicitly call back to that in the text though.